### PR TITLE
Server cookies synchronization with client (part of #1137, fix #1280 and fix #962)

### LINKED ISSUE #962: Set client cookie from the `set-cookie` header of `fetch` request
Similar problem was fixed in the #956 PR for xhr requests

### LINKED ISSUE #1280: Cookies from image response headers do not synchronize with client
Site https://allabout.co.jp/.
Error text:
`Uncaught TypeError: i.split is not a function`
Server for reproducing:
```javascript
require('http').createServer(function (req, res) {
    if (req.url === '/') {
        res.writeHead(200, { 'Content-Type': 'text/html' });
        res.end(`
            <img src="/img">
            <script>
                document.querySelector('img').addEventListener('load', function () {
                    var cookies = document.cookie.split('; ');
                    var someCookie;

                    for (var i = 0; i < cookies.length; i++) {
                        var parsedCookie = cookies[i].split('=');

                        if (parsedCookie[0] === 'some')
                            someCookie = cookies[i];
                    }

                    alert(someCookie.toLowerCase());
                });
            </script>
        `);
    } else if (req.url === '/img') {
        res.writeHead(200, {
            'Content-Type': 'image/gif',
            'Set-Cookie': 'some=cookie'
        });
        res.end(new Buffer('R0lGODlhAQABAID/AMDAwAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==', 'base64'));
    } else {
        res.end();
    }
}).listen(1993);
```

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

File: src/client/sandbox/fetch.js [hunk 1]
```diff
@@ -121,7 +123,12 @@ export default class FetchSandbox extends SandboxBase {
 
             sandbox.emit(sandbox.FETCH_REQUEST_SENT_EVENT, fetchPromise);
 
-            return fetchPromise;
+            return fetchPromise
+                .then(response => {
+                    sandbox.cookieSandbox.syncServerCookie();
```
Comment (miherlosev) [inline src/client/sandbox/fetch.js]: I guess we should perform cookie synchronization before sending fetch request (before call `nativeMethods.fetch.apply(this, args);`)

File: src/utils/cookie.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// -------------------------------------------------------------
+// WARNING: this file is used by both the client and the server.
+// Do not use any browser or node-specific API!
+// -------------------------------------------------------------
+
+import trim from './string-trim';
+
+const SERVER_SYNCHRONIZATION_FLAG = 's';
```
Comment (miherlosev) [inline src/utils/cookie.js]:
```js
const SYNCHRONIZATION_TYPE = {
   SERVER: 's',
  CLIENT: 'c'
}
```

File: src/utils/cookie.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+        if (parsedCookie)
+            parsedCookies.push(parsedCookie);
+    }
+
+    return parsedCookies;
+}
+
+export function formatSyncCookie (cookie) {
+    const flag         = cookie.isServerSync ? SERVER_SYNCHRONIZATION_FLAG : CLIENT_SYNCHRONIZATION_FLAG;
```
Comment (miherlosev) [inline src/utils/cookie.js]: `const synchonizationType = ...`

File: src/utils/cookie.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        domain:       decodeURIComponent(parsedKey[3]),
+        path:         decodeURIComponent(parsedKey[4]),
+        expires:      parsedKey[5] ? new Date(parseInt(parsedKey[5], 10)) : 'Infinity',
+        lastAccessed: new Date(parseInt(parsedKey[6], 10)),
+        value:        parsedCookie.join('='),
+        syncKey:      key
+    };
+}
+
+export function isObsoleteSyncCookie (currentCookie, newCookie) {
```
Comment (miherlosev) [inline src/utils/cookie.js]: `isOutdatedCookieSync`

File: src/utils/cookie.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+           newCookie.path === currentCookie.path &&
+           newCookie.lastAccessed > currentCookie.lastAccessed;
+}
+
+export function generateDeleteSyncCookieStr (cookie) {
+    return cookie.syncKey + '=;path=/;expires=Thu, 01 Jan 1970 00:00:01 GMT';
+}
+
+export function processServerCookie (ctx, parsedCookies) {
+    parsedCookies = parsedCookies.filter(cookie => !cookie.httpOnly);
```
Comment (miherlosev) [inline src/utils/cookie.js]: Store array's methods to the separate variables at the start of the file

File: src/client/sandbox/cookie/index.js [hunk 6]
```diff
@@ -127,7 +130,7 @@ export default class CookieSandbox extends SandboxBase {
 
         // NOTE: First, update our client cookies cache with a client-validated cookie string,
         // so that sync code can immediately access cookies.
-        const parsedCookie = cookieUtils.parse(value);
+        const parsedCookie = typeof value === 'string' ? cookieUtils.parse(value) : value;
```
Comment (miherlosev) [inline src/client/sandbox/cookie/index.js]: Add test for this change

File: src/utils/cookie.js (hunk 5 above)
Comment (LavrovArtem) [reply src/utils/cookie.js]: This function is called only on server side.

File: src/client/sandbox/cookie/index.js (hunk 6 above)
Comment (LavrovArtem) [reply src/client/sandbox/cookie/index.js]: All client tests which test the server synchronization go through this condition.

File: src/utils/cookie.js (hunk 5 above)
Comment (miherlosev) [reply src/utils/cookie.js]: Then it confused. The header of this file contains comment that this code is shared between client and server.
I propose to move this function to the separate file and require `src/utils/cookie.js`

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/cookie/index.js [hunk 7]
```diff
@@ -162,4 +165,17 @@ export default class CookieSandbox extends SandboxBase {
 
         return value;
     }
+
+    syncServerCookie () {
+        const cookies       = nativeMethods.documentCookieGetter.call(this.document);
+        const parsedCookies = parseClientSyncCookieStr(cookies);
+
+        for (const parsedCookie of parsedCookies) {
+            if (settings.get().sessionId === parsedCookie.sid && parsedCookie.isServerSync) {
```
Comment (churkin) [inline src/client/sandbox/cookie/index.js]: define `settings.get().sessionId` out of the loop

**Comment by LavrovArtem** [comment]
ping @churkin @miherlosev

File: src/utils/cookie.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    return newCookie.isServerSync === currentCookie.isServerSync &&
+           newCookie.sid === currentCookie.sid &&
+           newCookie.key === currentCookie.key &&
+           newCookie.domain === currentCookie.domain &&
+           newCookie.path === currentCookie.path &&
+           newCookie.lastAccessed > currentCookie.lastAccessed;
+}
+
+export function generateDeleteSyncCookieStr (cookie) {
+    return cookie.syncKey + '=;path=/;expires=Thu, 01 Jan 1970 00:00:01 GMT';
```
Comment (miherlosev) [inline src/utils/cookie.js]: `return cookie.syncKey + CLEAR_COOKIE_VALUE_STR`;

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
Before merge I will run health monitor on this PR

**Pull request merged**